### PR TITLE
Explore: Fix rendering of react query editors

### LINKED ISSUE #20923: Explore: Support react QueryEditor component in explore
Currently, explore only checks for ExporeQueryField or the angular query editor. It should also check/support the default QueryEditor interface.

### LINKED ISSUE #24564: explore: stackdriver data source not loading query editor
**What happened**:

Opened Explore and switched to a Stackdriver data source. No query editor loaded and got the following JavaScript error in the console - from `plugin_component.ts:265`:

```js
Plugin component error 
{message: "Failed to find exported plugin component for query-ctrl-stackdriver"}
message: "Failed to find exported plugin component for query-ctrl-stackdriver"
__proto__: Object
```

**What you expected to happen**:

That the query editor would load in Explore.

**How to reproduce it (as minimally and precisely as possible)**:
1. Set up a Stackdriver data source (provisioning files for data sources and dashboards can be found [in our private plugin provisioning repo](https://github.com/grafana/plugin-provisioning/tree/master/plugins/stackdriver)
2. Open Explore and try to switch to Stackdriver:

image

**Anything else we need to know?**:

**Environment**:
- Grafana version: latest master
- Data source type & version: Stackdriver
- OS Grafana is installed on: Ubuntu 16.10
- User OS & Browser: Ubuntu 16.10 and Chrome
- Grafana plugins:
- Others:

### PR DESCRIPTION
**Pull request by ivanahuckova**
**What this PR does / why we need it**:
Currently in Explore we are not checking if data source has react QueryEditor. This PR fixes it by checking:
- if data source has explore fields -> render explore fields
- if data source has react query editor -> render react query editor
- if data source doesn't have explore fields and query editor -> render angular editor

Before:
image

After:
image

**Which issue(s) this PR fixes**:

Fixes #24564
Fixes: #20923

### CODE REVIEW

**Review by Estrax — APPROVED** [review]
LGTM

**Comment by Estrax** [comment]
Can we add a test checking that it renders proper editors? Also, are we sure that all the tracing editors are going to be handled by `ExploreQueryField`?

File: public/app/features/explore/QueryRow.tsx [hunk 1]
```diff
@@ -155,8 +163,20 @@ export class QueryRow extends PureComponent<QueryRowProps, QueryRowState> {
                 exploreMode={mode}
                 exploreId={exploreId}
               />
-            ) : (
+            )}
+            {!ExploreQueryField && QueryEditor && (
```
Comment (aocenas) [inline public/app/features/explore/QueryRow.tsx:167]: Can we encapsulate this decision making a bit more. Right now there are 2 section where we decide what to render `setExploreQueryField` method and then here in render dispersed with the markup. I think we can refactor it so there is one place where the decision is made.

File: public/app/features/explore/QueryRow.tsx [hunk 2]
```diff
@@ -101,6 +101,21 @@ export class QueryRow extends PureComponent<QueryRowProps, QueryRowState> {
     this.setState({ textEditModeEnabled: !this.state.textEditModeEnabled });
   };
 
+  setExploreQueryField = () => {
```
Comment (aocenas) [inline public/app/features/explore/QueryRow.tsx:104]: Some test for the new behavior would be nice.

**Comment by ivanahuckova** [comment]
@aocenas @Estrax I have added test coverage and refactored solution. Can you check it out? 🙂

File: public/app/features/explore/QueryRow.tsx [hunk 3]
```diff
@@ -101,16 +101,23 @@ export class QueryRow extends PureComponent<QueryRowProps, QueryRowState> {
     this.setState({ textEditModeEnabled: !this.state.textEditModeEnabled });
   };
 
-  updateLogsHighlights = debounce((value: DataQuery) => {
```
Comment (ivanahuckova) [inline public/app/features/explore/QueryRow.tsx:104]: This is the same as on line 16. Not sure why it marked as deleted and then added.🤔

**Review by aocenas — APPROVED** [review]
Nice work

**Comment by torkelo** [comment]
yay!!! ❤️

**Pull request merged**